### PR TITLE
add EdgeAwareAugmentation for structure-preserving augmentations

### PR DESCRIPTION
**Pull request by janvis11**
## 📝 Description

Implements edge-aware augmentation that uses edge/structure priors to modulate augmentation strength. This preserves object boundaries and fine-grained spatial structure during augmentation, particularly useful for VLM grounding tasks and VLA/robotics pipelines.

**Fixes/Relates to:** #3676 

## 🛠️ Changes Made
- New `EdgeAwareAugmentation` class in `kornia/augmentation/_2d/intensity/edge_aware.py`
- Supports Sobel and Canny edge detectors
- Two modulation modes: "soft" (continuous) and "hard" (binary)
- Configurable `edge_weight` parameter to control preservation strength
- Can wrap any base augmentation module

---

## 🧪 How Was This Tested?
Comprehensive test suite added in `tests/augmentation/test_edge_aware.py`:
- Smoke tests with various parameter combinations
- Exception tests for invalid parameters
- Cardinality tests
- Edge preservation verification
- Mode comparison tests

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## Usage Example

```python
from kornia.augmentation import EdgeAwareAugmentation, RandomGaussianBlur

base_aug = RandomGaussianBlur(kernel_size=(3, 3), sigma=(0.1, 2.0), p=1.0)
aug = EdgeAwareAugmentation(
    base_aug,
    edge_detector="sobel",
    mode="soft",
    edge_weight=0.3,
    p=1.0
)
output = aug(input_tensor) 

---

### CODE REVIEW

File: kornia/augmentation/_2d/intensity/edge_aware.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    augmentation strength in high-edge regions while allowing stronger augmentation in smooth regions.
+    Particularly useful for VLM grounding tasks and VLA/robotics pipelines where spatial structure matters.
+
+    .. image:: _static/img/EdgeAwareAugmentation.png
+
+    Args:
+        base_aug: The base augmentation module to apply (e.g., RandomBrightness, RandomGaussianBlur).
+        edge_detector: Edge detection method, either "sobel" or "canny".
+        mode: Modulation mode, either "soft" (continuous weighting) or "hard" (binary masking).
+        edge_weight: Weight for edge modulation. Higher values preserve edges more strongly.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/edge_aware.py]: I assume the edge_weight is bounded to [0, 1]? Please make the docs more accurate.

So does the soft/hard. I need to read the code to understand what it does.

Some validations are needed in the __init__

File: kornia/augmentation/_2d/intensity/edge_aware.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+from kornia.augmentation._2d.intensity.base import IntensityAugmentationBase2D
+from kornia.filters import canny, sobel
+
+
+class EdgeAwareAugmentation(IntensityAugmentationBase2D):
+    r"""Edge-aware augmentation that uses edge/structure priors to modulate augmentation strength.
+
+    This augmentation preserves object boundaries and fine-grained spatial structure by reducing
+    augmentation strength in high-edge regions while allowing stronger augmentation in smooth regions.
+    Particularly useful for VLM grounding tasks and VLA/robotics pipelines where spatial structure matters.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/intensity/edge_aware.py]:
```suggestion
```
This is unnecessary if it is true.

**Review by shijianjian — COMMENTED** [review]
Is the augmentation related to a paper? If so, please add citations to it. If not, please attach some docs to this PR.

**Review by shijianjian — CHANGES_REQUESTED** [review]
The current implementation calls `self.base_aug(input)` inside `apply_transform()`, which bypasses Kornia’s normal parameter sampling / replay lifecycle. This can lead to desynchronization in `AugmentationSequential`, especially for multi-input pipelines (e.g. masks / boxes / keypoints), and makes `same_on_batch` semantics misleading.

I would strongly suggest applying the wrapped transform using pre-generated params (or restricting `base_aug` to deterministic modules) instead of invoking its `forward()` inside `apply_transform()`.

Also, edge-map normalization currently uses global batch min/max, which couples samples in the batch. This should likely be normalized per sample. The `p` seems also redundant in the current implementation. 

I feel like it will need a bigger revision. And I am not sure the current design fits into the augmentation module well. Many factors are not considered.

**Review by edgarriba — COMMENTED** [review]
Thanks for the effort here — the tests and docs were well put together, and CI is green. Unfortunately the core design conflicts with kornia's augmentation architecture: calling the wrapped augmentation's `forward()` inside `apply_transform` bypasses the parameter sampling/replay lifecycle, which breaks `AugmentationSequential` sync for masks/boxes/keypoints and makes `same_on_batch` misleading; the edge-map normalization also uses global batch min/max, coupling samples; and accepting any `nn.Module` as `base_aug` means wrapping a geometric aug blends spatially misaligned images.

Since the requested changes from April went unanswered, I'm inclined to close this to keep the queue healthy. If you'd like to pursue it, let's continue on #3676 first — a reference paper would help a lot — and agree on a design that pre-generates params and composes safely with `AugmentationSequential`. A fresh PR after that discussion would be very welcome.

**Comment by edgarriba** [comment]
Closing to keep the queue healthy — three months without response to the requested changes, and the core design (calling the wrapped aug's forward() inside apply_transform) conflicts with the augmentation lifecycle as detailed in the review. Please continue the design discussion on #3676 with a reference paper; a redesigned PR is welcome after that.
